### ISSUE TITLE
Improve performance of dask.array.shape

### ISSUE THREAD

**Issue by Illviljan** [dask/dask#7019]
**What happened**:
`dask.array.ndim` is a very common call in xarray. `ndim` is dependent on `dask.array.shape` that loops through the chunks and calls `cached_cumsum`, this appears to be rather expensive.

**What you expected to happen**:
Since chunks doesn't change _all_ the time I expected `shape` to be cached and very fast.

**Minimal Complete Verifiable Example**:
You'll probably find worse offenders if you don't have the changes from https://github.com/pydata/xarray/pull/4740 but if you profile and check what's going on during the interpolation you should see it:
```python
import xarray as xr
import numpy as np
import dask.array as da

a = np.arange(0, 2000)
b = np.core.defchararray.add("long_variable_name", a.astype(str))
coords = dict(time=da.array([0, 1]))
data_vars = dict()
for v in b:
    data_vars[v] = xr.DataArray(
        name=v,
        data=da.array([3, 4]),
        dims=["time"],
        coords=coords
    )
ds0 = xr.Dataset(data_vars)
ds0 = ds0.interp(
    time=da.array([0, 0.5, 1]),
    assume_sorted=True,
    kwargs=dict(fill_value=None),
)
```

**Anything else we need to know?**:
Current implementation in the Array class:
```python
    @property
    def shape(self):
        return tuple(cached_cumsum(c, initial_zero=True)[-1] for c in self.chunks)
```

I was able to halve the time of `ndim` by caching the shape like this:
```python
    @property
    def shape(self):
        chunks = self.chunks
        if self._cache.get("shape_chunks_old", None) != chunks:
            self._cache["shape_chunks_old"] = chunks
            self._cache["shape"] = tuple(cached_cumsum(c, initial_zero=True)[-1] for c in chunks)

        return self._cache["shape"]
```

Related issues:
* https://github.com/pydata/xarray/issues/4739
* https://github.com/pydata/xarray/pull/4740

**Environment**:
<details><summary>Output of <tt>xr.show_versions()</tt></summary>

xr.show_versions()

INSTALLED VERSIONS
------------------
commit: None
python: 3.8.5 (default, Sep  3 2020, 21:29:08) [MSC v.1916 64 bit (AMD64)]
python-bits: 64
OS: Windows
OS-release: 10
libhdf5: 1.10.4
libnetcdf: None

xarray: 0.16.2
pandas: 1.1.5
numpy: 1.17.5
scipy: 1.4.1
netCDF4: None
pydap: None
h5netcdf: None
h5py: 2.10.0
Nio: None
zarr: None
cftime: None
nc_time_axis: None
PseudoNetCDF: None
rasterio: None
cfgrib: None
iris: None
bottleneck: 1.3.2
dask: 2020.12.0
distributed: 2020.12.0
matplotlib: 3.3.2
cartopy: None
seaborn: 0.11.1
numbagg: None
pint: None
setuptools: 51.0.0.post20201207
pip: 20.3.3
conda: 4.9.2
pytest: 6.2.1
IPython: 7.19.0
sphinx: 3.4.0

</details>

**Comment by shoyer** [comment]
On Python 3.8+, this could use [`functools.cached_property`](https://docs.python.org/3/library/functools.html#functools.cached_property)

**Comment by Illviljan** [comment]
I gave `functools.cached_property` a try and with that implementation it reduces the time by 1/3. It's looking quite good to be honest, and it can very easily be added to other popular properties as well significantly lowering total calculation time. Xarray could probably use the same thing as now similar calculations are creeping up as the bottleneck in my test cases.

One caveat about the cached_property is that if the chunks updates for some reason it will eagerly trigger a shape calculation as well even though I did not want to know the shape then. I haven't managed to make it lazier without having to rethink cached_property though.
